keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -237,6 +195,91 @@ class ResourceActuator(
     }
   }
 
+  /**
+   * [VersionedArtifactProvider] is a special [resource] sub-type. When a veto response sets
+   * [VetoResponse.vetoArtifact], the resource under evaluation is of type
+   * [VersionedArtifactProvider], and the version has never before been deployed successfully to an environment,
+   * disallow the desired artifact version from being deployed to
+   * the environment containing [resource]. This ensures that the environment will be fully restored to
+   * a prior good-state.
+   *
+   * If the version has previously been deployed successfully do not veto the artifact because
+   * we do not want to veto the artifact in the case of a bad config change or other downstream problem.
+   * Rolling back most likely will not help in these cases and it will probably cause confusion.
```

Comment:
I'm fairly confused here. Is the `VetoResponse` in reference to the same artifact version that `desired` is referencing? That implies that something has already decided this version should be vetoed but we're now deciding whether to honor that decision? 

Or is the version `desired` references the one we're considering rolling back to?

Reply:
Yes, you're correct. The veto response doesn't specify the version, just whether the artifact version that is present should be vetoed. I agree that ideally the logic for vetoing an artifact should belong in the actual veto! Unfortunately, I don't feel we have enough data in the veto right now to make the right call (I describe this further in the "thoughts" section of my PR description). So, the way I implemented this was to decide whether or not to honor that veto in the `ResourceActuator`.

Reply:
Updated the method description to be a bit more detailed